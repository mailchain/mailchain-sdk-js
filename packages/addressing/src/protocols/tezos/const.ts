export enum Prefix {
	TZ1 = 'tz1',
	TZ2 = 'tz2',
	TZ3 = 'tz3',
	KT = 'KT',
	KT1 = 'KT1',

	EDSK2 = 'edsk2',
	SPSK = 'spsk',
	P2SK = 'p2sk',

	EDPK = 'edpk',
	SPPK = 'sppk',
	P2PK = 'p2pk',

	EDESK = 'edesk',
	SPESK = 'spesk',
	P2ESK = 'p2esk',

	EDSK = 'edsk',
	EDSIG = 'edsig',
	SPSIG = 'spsig',
	P2SIG = 'p2sig',
	SIG = 'sig',
	EXPR = 'expr',
}

export const prefix = {
	[Prefix.TZ1]: new Uint8Array([6, 161, 159]),
	[Prefix.TZ2]: new Uint8Array([6, 161, 161]),
	[Prefix.TZ3]: new Uint8Array([6, 161, 164]),
	[Prefix.KT]: new Uint8Array([2, 90, 121]),
	[Prefix.KT1]: new Uint8Array([2, 90, 121]),

	[Prefix.EDSK]: new Uint8Array([43, 246, 78, 7]),
	[Prefix.EDSK2]: new Uint8Array([13, 15, 58, 7]),
	[Prefix.SPSK]: new Uint8Array([17, 162, 224, 201]),
	[Prefix.P2SK]: new Uint8Array([16, 81, 238, 189]),

	[Prefix.EDPK]: new Uint8Array([13, 15, 37, 217]),
	[Prefix.SPPK]: new Uint8Array([3, 254, 226, 86]),
	[Prefix.P2PK]: new Uint8Array([3, 178, 139, 127]),

	[Prefix.EDESK]: new Uint8Array([7, 90, 60, 179, 41]),
	[Prefix.SPESK]: new Uint8Array([0x09, 0xed, 0xf1, 0xae, 0x96]),
	[Prefix.P2ESK]: new Uint8Array([0x09, 0x30, 0x39, 0x73, 0xab]),

	[Prefix.EDSIG]: new Uint8Array([9, 245, 205, 134, 18]),
	[Prefix.SPSIG]: new Uint8Array([13, 115, 101, 19, 63]),
	[Prefix.P2SIG]: new Uint8Array([54, 240, 44, 52]),
	[Prefix.SIG]: new Uint8Array([4, 130, 43]),
	[Prefix.EXPR]: new Uint8Array([13, 44, 64, 27]),
};
