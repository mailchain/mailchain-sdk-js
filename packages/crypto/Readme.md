# Mailchain Crypto

Mailchain is a messaging protocol that lets users communicate across protocols. Using Mailchain you can send messages to any blockchain address on different protocols.

For full usage examples view the [developer docs](https://docs.mailchain.com).

## Installing

Using npm:

```bash
$ npm install @mailchain/crypto
```

Using yarn:

```bash
$ yarn add @mailchain/crypto
```

## Purpose

The crypto package is used by `@mailchain/sdk` to support securely registering blockchain addresses and sending messages via Mailchain.
