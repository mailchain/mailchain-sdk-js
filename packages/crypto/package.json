{
	"name": "@mailchain/crypto",
	"version": "0.31.0",
	"license": "Apache-2.0",
	"dependencies": {
		"@ethersproject/hash": "5.7.0",
		"@ethersproject/signing-key": "5.6.2",
		"@mailchain/encoding": "0.31.0",
		"@noble/curves": "^0.8.2",
		"@noble/hashes": "^1.3.0",
		"@scure/bip39": "^1.2.0",
		"bn.js": "^5.2.1",
		"ed2curve": "^0.3.0",
		"secp256k1": "^4.0.2",
		"tweetnacl": "^1.0.3"
	},
	"devDependencies": {
		"@types/bn.js": "^5.1.0",
		"@types/ed2curve": "^0.2.3",
		"@types/secp256k1": "^4.0.2"
	},
	"repository": {
		"type": "git",
		"url": "https://github.com/mailchain/mailchain-sdk-js.git",
		"directory": "crypto"
	},
	"bugs": {
		"url": "https://github.com/mailchain/mailchain-sdk-js/issues"
	},
	"engines": {
		"node": ">=15.0.0"
	},
	"scripts": {
		"build": "run -T tsup",
		"test": "run -T jest",
		"lint:test": "yarn run -T eslint --ext js,ts,tsx src",
		"lint:fix": "yarn run -T eslint --fix --ext js,ts,tsx src"
	}
}
