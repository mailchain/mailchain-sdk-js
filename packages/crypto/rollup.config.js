import { getBundlingConfigs } from '../library.rollup.config';
import pkg from './package.json';

export default getBundlingConfigs('crypto', pkg, { preserveModules: true });
