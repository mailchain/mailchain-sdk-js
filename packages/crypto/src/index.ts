export { ED25519ExtendedPrivateKey, ED25519PrivateKey, ED25519PublicKey, deriveHardenedKey } from './ed25519';
export { SECP256K1PrivateKey, SECP256K1PublicKey } from './secp256k1';
export * from './keys';
export * from './multikey';
export * from './public';
export * from './private';
export * from './rand';
export * from './hd';
export * from './cipher';
export * from './errors';
