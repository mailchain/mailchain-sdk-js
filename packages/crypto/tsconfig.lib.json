{
	"extends": "./tsconfig.json",
	"compilerOptions": {
		"target": "ES2017",
		"noEmit": false,
		"declaration": true,
		"incremental": false,
		"lib": ["es2017", "es7", "es6", "dom"]
	},
	"include": ["**/*.ts"],
	"exclude": ["jest.config.cjs", "**/*.test.ts"]
}
