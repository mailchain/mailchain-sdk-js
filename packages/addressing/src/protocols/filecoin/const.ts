import { SECP256K1PrivateKey } from '@mailchain/crypto';
import { decodeHex } from '@mailchain/encoding';

// Alice
export const AliceFilStr = 'f410faiikgixcdd7alrsbpunjlen4etkcarbfbghn7wi';
export const AliceFilEthAddressStr = '0x0210a322e218fe05c6417d1a9591bc24d4204425';
export const AliceFilEthPrivateKeyStr = '6c115551accbd74b46401a531af1dd0890034c18137dddab52a53c626e75b051';
export const AliceFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(AliceFilEthPrivateKeyStr));

// Bob
export const BobFilAddressStr = 'f410f5blfdi7tmto6o2aa2up4vwtdm7qcnaiqpeazuwy';
export const BobFilEthAddressStr = '0xe85651a3f364dde76800d51fcada6367e0268110';
export const BobFilEthPrivateKeyStr = '2c6b29d5545e5804b7fb7e65404782e57dddea84cbcec9b2586add6403808965';
export const BobFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(BobFilEthPrivateKeyStr));

export enum FilPrefix {
	F4ETHEREUM = 'f410',
	T4ETHEREUM = 't410',
}
