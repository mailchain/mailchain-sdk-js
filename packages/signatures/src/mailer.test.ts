import {
	AliceED25519PublicKey,
	BobED25519PublicKey,
	AliceED25519PrivateKey,
	BobED25519PrivateKey,
} from '@mailchain/crypto/ed25519/test.const';
import { AliceSECP256K1PublicKey, AliceSECP256K1PrivateKey } from '@mailchain/crypto/secp256k1/test.const';
import { ErrorUnsupportedKey } from '@mailchain/crypto';
import {
	createMailerProof,
	MailerProof,
	MailerProofParams,
	createMailerProofSigningData,
	verifyMailerProof,
} from './mailer';

const aliceMailerProof = {
	params: {
		authorContentSignature: Uint8Array.from([5, 6, 7, 8]),
		expires: new Date('2022-06-06'),
		mailerMessagingKey: BobED25519PublicKey,
	},
	signature: Uint8Array.from([
		1, 248, 81, 21, 91, 149, 47, 249, 46, 65, 40, 134, 3, 190, 218, 174, 82, 137, 251, 192, 231, 8, 229, 28, 105,
		119, 220, 38, 241, 57, 43, 202, 229, 218, 65, 246, 184, 235, 119, 36, 52, 5, 229, 247, 60, 101, 184, 217, 43,
		51, 117, 30, 197, 222, 236, 96, 253, 166, 217, 198, 241, 7, 242, 15,
	]),
	version: '1.0',
} as MailerProof;

const bobMailerProof = {
	params: {
		authorContentSignature: Uint8Array.from([5, 6, 7, 8]),
		expires: new Date('2022-06-06'),
		mailerMessagingKey: AliceED25519PublicKey,
	},
	signature: Uint8Array.from([
		255, 187, 63, 176, 181, 113, 75, 167, 184, 98, 33, 229, 194, 55, 116, 79, 39, 244, 40, 197, 105, 70, 235, 45,
		251, 228, 131, 215, 187, 15, 101, 59, 197, 153, 209, 177, 37, 180, 218, 37, 57, 209, 48, 186, 121, 117, 248, 97,
		12, 46, 41, 99, 68, 35, 7, 67, 124, 63, 81, 225, 77, 45, 13, 13,
	]),
	version: '1.0',
} as MailerProof;
describe('createMailerProofParamsForSigning()', () => {
	const tests = [
		{
			name: `version-1.0`,
			data: {
				authorContentSignature: new Uint8Array([0x05, 0x06, 0x07, 0x08]),
				expires: new Date('2022-06-06'),
				mailerMessagingKey: BobED25519PublicKey,
			} as MailerProofParams,
			version: '1.0',
		},
	];
	test.each(tests)('$name', (test) => {
		const actual = createMailerProofSigningData(test.data, test.version);
		expect(actual).toMatchSnapshot();
	});
});

describe('verifyMailerProof()', () => {
	const tests = [
		{
			name: `valid-sig-ed25519-alice`,
			args: {
				key: AliceED25519PublicKey,
				mailerProof: aliceMailerProof,
			},
			expected: true,
			shouldThrow: false,
		},
		{
			name: `valid-sig-ed25519-bob`,
			args: {
				key: BobED25519PublicKey,
				mailerProof: bobMailerProof,
			},
			expected: true,
			shouldThrow: false,
		},
		{
			name: `incorrect-sig-ed25519-alice`,
			args: {
				key: AliceED25519PublicKey,
				mailerProof: {
					...aliceMailerProof,
					signature: new Uint8Array([
						0x50, 0x2a, 0x4d, 0x32, 0x20, 0x94, 0x45, 0xda, 0xaf, 0x82, 0xc2, 0xfd, 0x73, 0xa9, 0x83, 0xeb,
						0xd, 0xfb, 0x45, 0x92, 0x9b, 0x90, 0x46, 0xbb, 0x7c, 0x1d, 0xcd, 0x6d, 0xa7, 0x87, 0x8d, 0x6f,
						0x4c, 0x45, 0x53, 0xd9, 0x2, 0xf8, 0xa6, 0x12, 0xae, 0x6f, 0x46, 0xd3, 0x4f, 0x75, 0x59, 0x30,
						0xa8, 0xd1, 0xaa, 0xcb, 0xd, 0xf0, 0x5, 0xff, 0xc1, 0xb4, 0x7e, 0xe7, 0xe8, 0x3d, 0x4f, 0xc,
					]),
				},
			},
			expected: false,
			shouldThrow: false,
		},
		{
			name: `incorrect-sig-ed25519-bob`,
			args: {
				key: BobED25519PublicKey,
				mailerProof: {
					...bobMailerProof,
					signature: new Uint8Array([
						0xfe, 0xac, 0xe1, 0x62, 0xd7, 0x82, 0xae, 0xd, 0xff, 0x85, 0x9f, 0xe2, 0xc, 0x2a, 0xdd, 0xbe,
						0xf5, 0xbf, 0x3c, 0xd4, 0x8f, 0x49, 0x5c, 0x45, 0x97, 0x7f, 0x87, 0x4f, 0xf5, 0xcf, 0x50, 0x67,
						0xac, 0x20, 0x88, 0xf9, 0xc, 0x62, 0x68, 0xb9, 0xcf, 0x51, 0x7c, 0x51, 0xc4, 0x93, 0xae, 0xdb,
						0x5b, 0x10, 0xfd, 0x53, 0x64, 0x90, 0x87, 0x83, 0x9f, 0x87, 0xe8, 0x8a, 0x32, 0x38, 0x84, 0xd,
					]),
				},
			},
			expected: false,
			shouldThrow: false,
		},
		{
			name: `unsupported-key`,
			args: {
				key: AliceSECP256K1PublicKey,
				mailerProof: aliceMailerProof,
			},
			expected: false,
			shouldThrow: new ErrorUnsupportedKey('secp256k1'),
		},
	];
	test.each(tests)('$name', async (test) => {
		const target = verifyMailerProof(test.args.key, test.args.mailerProof);
		if (test.shouldThrow) {
			expect.assertions(1);
			return target.catch((e) => expect(e).toEqual(test.shouldThrow));
		}
		return target.then((actual) => {
			expect(actual).toEqual(test.expected);
		});
	});
});

describe('createMailerProof()', () => {
	const tests = [
		{
			name: `ed25519-alice`,
			args: {
				key: AliceED25519PrivateKey,
				message: {
					expires: new Date('2022-06-06'),
					mailerMessagingKey: BobED25519PublicKey,
					authorContentSignature: new Uint8Array([0x05, 0x06, 0x07, 0x08]),
				} as MailerProofParams,
				version: '1.0',
			},
			expected: aliceMailerProof,
			shouldThrow: false,
		},
		{
			name: `ed25519-bob`,
			args: {
				key: BobED25519PrivateKey,
				message: {
					authorContentSignature: new Uint8Array([0x05, 0x06, 0x07, 0x08]),
					expires: new Date('2022-06-06'),
					mailerMessagingKey: AliceED25519PublicKey,
				} as MailerProofParams,
				version: '1.0',
			},
			expected: bobMailerProof,
			shouldThrow: false,
		},
		{
			name: `unsupported-key`,
			args: {
				key: AliceSECP256K1PrivateKey,
				message: {
					authorContentSignature: new Uint8Array([0x05, 0x06, 0x07, 0x08]),
					expires: new Date('2022-06-06'),
					mailerMessagingKey: BobED25519PublicKey,
				} as MailerProofParams,
				version: '1.0',
			},
			expected: null,
			shouldThrow: new ErrorUnsupportedKey('secp256k1'),
		},
	];
	test.each(tests)('$name', async (test) => {
		const target = createMailerProof(test.args.key, test.args.message, test.args.version);
		if (test.shouldThrow) {
			expect.assertions(1);
			return target.catch((e) => expect(e).toEqual(test.shouldThrow));
		}
		return target.then((actual) => {
			expect(actual).toEqual(test.expected);
		});
	});
});
