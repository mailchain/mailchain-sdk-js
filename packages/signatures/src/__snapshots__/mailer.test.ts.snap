// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMailerProofParamsForSigning() version-1.0 1`] = `"{"authorContentSignature":"05060708","expires":1654473600,"mailerMessagingKey":"e22e322f8740c60172111ac8eadcdda2512f90d06d0e503ef189979a159bece1e8"}"`;
